### ISSUE TITLE
`noqa` comment stripped due to syntax error elsewhere in file

### ISSUE THREAD

**Issue by stinodego** [charliermarsh/ruff#4849]
Ruff indicates an unused noqa statement, when in fact the comment is fine. There is a syntax error in the function definition that triggers this, if the `*` is removed, everything is fine.

```python
def function(*):
    """Veeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeery long."""  # noqa: W505
```

This is on ruff version `0.0.270`.

**Comment by charliermarsh** [comment]
Thanks, I actually thought we added handling for this case at some point. Will revisit!

**Comment by stinodego** [comment]
> Thanks, I actually thought we added handling for this case at some point. Will revisit!

I know, I have reported something similar before:
https://github.com/charliermarsh/ruff/issues/2853

Not sure what caused this one to resurface.

**Comment by charliermarsh** [comment]
(I believe this regressed when we started generating fixes on every pass, regardless of whether they're applied.)

**Comment by charliermarsh** [comment]
@MichaReiser - Any opinion on the right behavior here? The issue is that if we hit a syntax error, we fail to identify the W505 violation in the example above, and so the `noqa` is considered unused and removed.

We could...

1. Avoid autofixing at all if we hit a syntax error.
2. Avoid autofixing RUF100 violations (unused suppression comments)  if we hit a syntax error.
3. Avoid _raising_ RUF100 violations if we hit a syntax error.

**Comment by MichaReiser** [comment]
@charliermarsh My preference is 3. The way I think about this is that rules should only run if their input is correct. What the input is, differs from rule to rule:

* Physical lines: always valid
* logical lines: valid tokens
* AST rules: parsable source tree
* RUF100: All configured rules ran

That's why I think RUF100 should not run if any of the other phases returned an `Err`.

**Comment by stinodego** [comment]
For what it's worth, I also believe 3 is the way to go. I don't like all the linting violations popping up in seemingly unrelated places while I'm writing some code.

**Comment by charliermarsh** [comment]
Sounds good, will fix today.
